### ISSUE TITLE
Support for @user Mentions

### ISSUE THREAD

**Issue by WillStrohl** [DNNCommunity/Dnn.CommunityForums#158]
## Is your feature request related to a problem?
When writing in a post, if we want to involve anyone else in the conversation, we have to use any number of manual methods to do this.  For example, if I want to bring in someone to help another end-user with a question I know nothing about, I'll need to reply saying I will get someone to help, then go to any number of other places (e.g., Slack, Facebook, Skype, etc.) to find and ask someone to look at the forum post.  

## Describe the solution you'd like  
Instead, we should begin implementing a tagging feature that allows us to tag end-users like we do here and/or in other places like Facebook, Slack, etc.  Typing `@` would allow us to begin typing the `Display Name` of another user, and we can mention them.   

When a person is mentioned, they will get an email stating that they were mentioned by `Display Name` and include a link to the post. If that same person is already subscribed to the forum thread, they should only get one email in total.  

I can see this quickly being something that someone finds a way to take advantage of.  So, it may be worth the effort to only allow a person to tag someone who they've become "friends" with on the website.  This way, you can't just tag someone just because to abuse the feature.  For example, if the forum is being used on a corporate website, you don't want an end-user being able to tag the CEO in many situations.  :)  

## Describe alternatives you've considered  
None yet.  

## Additional context  
None yet, but this could open things up for also tagging forum posts later.

**Comment by Timo-Breumelhof** [comment]
Yes please.. :-)

**Comment by Timo-Breumelhof** [comment]
About who one can tag, what about a setting in the users forum settings where you can set if and by whom you want to be tagged? Options:
- Everyone
- Friends
- Forum admins
- No taging

**Comment by WillStrohl** [comment]
> About who one can tag, what about a setting in the users forum settings where you can set if and by whom you want to be tagged? Options:
> 
> * Everyone
> * Friends
> * Forum admins
> * No tagging

I think this is a great idea.  Based on my earlier brainstorming, do you agree that the default should be "Friends"?

**Comment by Timo-Breumelhof** [comment]
Ideally I would say a setting for what the default is? But I would start with Everyone I think.

**Comment by WillStrohl** [comment]
Well, there are two obvious paths to follow based on how the community usually sees these kinds of updates in general:  

1. The default should just be no change, which in this case would mean anyone who wants tagging would need to turn it on after upgrade.  (However, this runs the risk of site owners never discovering or using this feature.)  
2. Have a default that exposes the feature to everyone, likely causing a surge in activity as people get excited about and use the new tagging feature.  Thinking with my security hat on, this should default to Friends, but that's the same as "No Tagging" on most websites, until people begin friending each other. So, I guess Everyone might make sense, and they then lock it down as they see fit.  

Thoughts?

**Comment by Timo-Breumelhof** [comment]
@WillStrohl ,

1. I think we should just turn it on and communicate that you can turn it off. I wonder how many users will actually turn it off as it's a cool feature.
2. I agree. If we limit it to only friends it will most likely not make any impact.
My thought was actually more to only have an on/  off switch and leave the rest to the user in their own forum settings (not for the admin but for a forum user)

Extra: I think the most useful option would actually be that in a reply the name of the user you reply to automatically turns into a mention.. :-)

**Comment by WillStrohl** [comment]
I like your ideas, @Timo-Breumelhof ...  This is kind of the Facebook model as well, so it will be just as familiar to turn this feature on/off.  🤔

**Comment by Timo-Breumelhof** [comment]
I have been thinking about this a bit more and also spoke to @donker ..

Would this be an interesting core feature?
Also it's the implementation in DNN social reusable maybe?

Let's say we would have a core API for this.

I'm thinking:
A module passes
- a list of "preferred user ids" (in our case participant in an thread for instance)
- the filter being the typed characters after the @
-  a maximum number of users
B. The API returns a list of users with a preference for the passed ids

I think something this can be useful for other modules too.
I didn't research yet how this is implemented in DNN Social and if it could be re-used.

Thoughts? 
@WillStrohl @johnhenley @mitchelsellers @skamphuis to name a few ;-)

**Comment by valadas** [comment]
From a core perspective, if that is the way it ends up being done, then it would have to be an opt-in feature. This could be used to easily list all valid users.

**Comment by Timo-Breumelhof** [comment]
> From a core perspective, if that is the way it ends up being done, then it would have to be an opt-in feature. This could be used to easily list all valid users.

Yes, IMO only available for registered users, but with only that in place there is not much difference with the user profile security of the current profile pages, right? (the API should only return a link to the user profile and the User Display Name IMO)
One other option could be to make the availability depend on the existence of a user profile page?

**Comment by WillStrohl** [comment]
I prefer this being a core feature, if at all possible.  This would make it a universal implementation that all module developers (and website owners) could benefit from.  This would be a great thing for DNN Platform.
